collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java

```diff
@@ -0,0 +1,705 @@
+        // shared across independent activities.
+        HttpContext localContext = new BasicHttpContext();
+
+        localContext.setAttribute(HttpClientContext.COOKIE_STORE, getCookieStore());
+        localContext.setAttribute(HttpClientContext.CREDS_PROVIDER, getCredentialsProvider());
+
+        return localContext;
+    }
+
+    /**
+     * Create an httpClient with connection timeouts and other parameters set.
+     * Save and reuse the connection manager across invocations (this is what
+     * requires synchronized access).
+     *
+     * @return HttpClient properly configured.
+     */
+    private synchronized HttpClient createHttpClient(int timeout) {
+        // configure connection
+        SocketConfig socketConfig = SocketConfig.copy(SocketConfig.DEFAULT).setSoTimeout(
+                2 * timeout)
+                .build();
+
+        // if possible, bias toward digest auth (may not be in 4.0 beta 2)
+        List<String> targetPreferredAuthSchemes = new ArrayList<>();
+        targetPreferredAuthSchemes.add(AuthSchemes.DIGEST);
+        targetPreferredAuthSchemes.add(AuthSchemes.BASIC);
+
+        RequestConfig requestConfig = RequestConfig.copy(RequestConfig.DEFAULT)
+                .setConnectTimeout(timeout)
+                // support authenticating
+                .setAuthenticationEnabled(true)
+                // support redirecting to handle http: => https: transition
+                .setRedirectsEnabled(true)
+                .setMaxRedirects(1)
+                .setCircularRedirectsAllowed(true)
+                .setTargetPreferredAuthSchemes(targetPreferredAuthSchemes)
+                .setCookieSpec(CookieSpecs.DEFAULT)
+                .build();
+
+        return HttpClientBuilder.create()
+                .setDefaultSocketConfig(socketConfig)
+                .setDefaultRequestConfig(requestConfig)
+                .build();
+
+    }
+
+    private void enablePreemptiveBasicAuth(
+            HttpContext localContext, String host) {
+        AuthCache ac = (AuthCache) localContext
+                .getAttribute(HttpClientContext.AUTH_CACHE);
+        HttpHost h = new HttpHost(host);
+        if (ac == null) {
+            ac = new BasicAuthCache();
+            localContext.setAttribute(HttpClientContext.AUTH_CACHE, ac);
+        }
+        List<AuthScope> asList = buildAuthScopes(host);
+        for (AuthScope a : asList) {
+            if (a.getScheme().equalsIgnoreCase(AuthSchemes.BASIC)) {
+                ac.put(h, new BasicScheme());
+            }
+        }
+    }
+
+    private List<AuthScope> buildAuthScopes(String host) {
+        List<AuthScope> asList = new ArrayList<>();
+
+        AuthScope a;
+        // allow digest auth on any port...
+        a = new AuthScope(host, -1, null, AuthSchemes.DIGEST);
+        asList.add(a);
+        // and allow basic auth on the standard TLS/SSL ports...
+        a = new AuthScope(host, 443, null, AuthSchemes.BASIC);
```

Comment:
Also, I think the instantiation of each `AuthScope` into the `a` variable is a bit redundant. All of this could be done inside the `add(object)` method instead making this code cleaner.

Reply:
👍 Agreed